Notes/Proposals/Explicit-partial-type-inference.md

```diff
@@ -0,0 +1,313 @@
+>
+> ...
+>
+> #### 12.6.3.2 The first phase
+> **If the method call has the form `M<Y₁...Yᵥ>(E₁ ... Eₓ)` for each `Yᵢ` which is not `_`, fix type variable `Xᵢ` to `Yᵢ`.**
+> For each of the method arguments `Eᵢ`:
+
+If we consider dynamic overload resolution, we modify compile-time checks to reflect the change.
+
+> 12.6.5 Compile-time checking of dynamic member invocation
+>
+> First, if F is a generic method and type arguments were provided, then those **which are not `_`** are substituted for the type parameters in the parameter list. However, if type arguments were not provided, no such substitution happens.
+
+
+There raises the question if we can go further and allow `_` to be nested 
+See the following example.
+
+```csharp
+TResult GetResult<TResult>() where TResult : IEnumerable<string> {}
+
+var results = GetResult<List<_>>(); // Just specifying wrapper implementation.
+```
+
+It would be most beneficial during working with any generic wrappers, where we are curious about the wrapper implementation, but not which will be inside the wrapper.
+
+This kind of wildcard needs more information about the relation between type arguments to become useful.
+More information can be extracted from generic type constraints, which can bring also more inference power to simple `_` inferred types.
+Unfortunately, using information about constraints in type inference causes breaking change, see the following example.
+
+```csharp
+void M(object) {}
+void M<T, U>(T t) where T : IEnumerable<U> {}
+
+M("foo");
+```
+
+One option would be to somehow differ between old inference and new improved inference like that.
+
+```csharp
+void M(object) {} // 1.
+void M<T, U>(U t) where T : IEnumerable<U> {} // 2.
+
+M<_, _>("foo"); // calls 2.
+M("foo"); // calls 1.
+```
+
+Because of this breaking change, we can use the constraints only with `_` inferred types, which is not presented in the current language version so it doesn't cause a breaking change.
+
+One can feel that `M<_,_>(...)` should be resolved in the same manner as `M(...)`. 
+This is a reasonable argument and we feel that it should follow it.
+For this reason, we would postpone this extended feature to future improvements where we would be able to use the constraints without causing a breaking change.
+This could be done by some tool that patches the old code.
+
+#### Object construction
+
+> TODO
+
+#### Variable declaration
+
+> TODO
+
+#### Casting
+
+> TODO
+
+## Alternatives
+
+We can choose different char or keywords for determining compiler inferred type argument. Other options consist of `*`, `var`, `?`, `<<whitespace>>`. Although, we think that `_` fits better because the same char is used in deconstruction assignments or patterns.
```

Comment:
Should probably also mention having _no token at all_ as an option, e.g. `Dictionary<,> dictionary = GetDictionary(key: "a", value: 1);`

Reply:
[Fixed](https://github.com/TomatorCZ/master-thesis/blob/30ae5cddd31428892ee99532e6a57bb7e9f4fb2d/Notes/Proposals/Explicit-partial-type-inference.md?plain=1#L364)